master/pkg/schemas/defaults.go

```diff
@@ -0,0 +1,229 @@
+//    schemas.Merge(&config.CheckpointStorage, cluster_default_checkpoint_storage)
+//
+//    // Define any remaining undefined values.
+//    schemas.FillDefaults(&config)
+//
+func FillDefaults(obj interface{}) {
+	vObj := reflect.ValueOf(obj)
+	// obj can't be a non-pointer, because it edits in-place.
+	if vObj.Kind() != reflect.Ptr {
+		panic("FillDefaults must be called on a pointer")
+	}
+	// obj can't be a nil pointer, because FillDefaults(nil) doesn't make any sense.
+	if vObj.IsZero() {
+		panic("FillDefaults must be called on a non-nil pointer")
+	}
+	// Enter the recursive default filling with no default bytes for the root object (which must
+	// already exist), and starting with the name of the object type.
+	name := fmt.Sprintf("%T", obj)
+	vObj.Elem().Set(fillDefaults(vObj.Elem(), nil, false, name))
+}
+
+func fillDefaults(obj reflect.Value, defaultBytes []byte, inPtr bool, name string) reflect.Value {
+	switch obj.Kind() {
+	case reflect.Interface:
+		if obj.IsZero() {
+			// This doesn't make any sense; we need a type.
+			panic("got a nil interface as the obj to FillDefaults into")
+		}
+		obj.Elem().Set(fillDefaults(obj.Elem(), defaultBytes, inPtr, name))
+
+	case reflect.Ptr:
+		if obj.IsZero() {
+			if defaultBytes == nil {
+				// Nil pointer with no defaultBytes means we are done recursing.
+				return obj
+			}
+			// Otherwise, since we have default bytes, allocate the new object.
+			obj = reflect.New(obj.Type().Elem())
+			// Fill the object with default bytes.
+			err := json.Unmarshal(defaultBytes, obj.Interface())
+			if err != nil {
+				panic(
+					fmt.Sprintf(
+						"failed to unmarshal defaultBytes into %T: %v",
+						obj.Interface(),
+						string(defaultBytes),
+					),
+				)
+			}
+			// We already consumed defaultBytes, so set it to nil when we recurse.
+			obj.Elem().Set(fillDefaults(obj.Elem(), nil, true, name))
+		} else {
+			// Recurse into the element inside the pointer.
+			obj.Elem().Set(fillDefaults(obj.Elem(), defaultBytes, true, name))
+		}
+
+	case reflect.Struct:
+		// Get a source of defaults from a Defaultable or Schema interface.
+		// Use Addr so that if the DefaultSource is defined on a struct pointer, it still works.
+		var defaultSource interface{}
+		if defaultable, ok := obj.Addr().Interface().(Defaultable); ok {
+			defaultSource = defaultable.DefaultSource()
+		} else if schema, ok := obj.Addr().Interface().(Schema); ok {
+			defaultSource = schema.ParsedSchema()
+		}
+
+		// Iterate through all the fields of the struct once, applying defaults.
+		for i := 0; i < obj.NumField(); i++ {
+			var fieldDefaultBytes []byte
+			if defaultSource != nil {
+				// Is there a default for this field's tag?
+				fieldDefaultBytes = findDefaultInSchema(defaultSource, obj.Type().Field(i))
+			}
+			fieldName := fmt.Sprintf("%v.%v", name, obj.Type().Field(i).Name)
+			// Recurse into the field.
+			fillDefaults(obj.Field(i).Addr(), fieldDefaultBytes, false, fieldName)
```

Comment:
Seems like a bug.  I'll add a test case to catch it, and fix it.

Reply:
I think this worked because there are certain kinds of values where we always alter the original value and return it, so I think this was accidentally totally valid.

In any case, I changed it to be explicitly correct.

Reply:
Ok, make sense.